File: Java/Ema/Core/src/main/java/com/refinitiv/ema/domain/directory/DirectoryServiceFilter.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2026 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.ema.domain.directory;

import com.refinitiv.ema.access.OmmInvalidUsageException;
import com.refinitiv.ema.rdm.EmaRdm;

/**
 * Base contract for an RDM Source Directory service filter.
 * <p>
 * A service filter encapsulates one portion of a Source Directory service definition,
 * such as service info, state, load, data, group, or link information.
 * Implementations expose their content as an EMA container type that can be encoded
 * into, or decoded from, a directory payload.
 *
 * @param <T> EMA container type used to represent the encoded filter payload,
 *            for example {@link com.refinitiv.ema.access.ElementList} or
 *            {@link com.refinitiv.ema.access.Map}
 */
public interface DirectoryServiceFilter<T>
{
    /**
     * Clears this service filter and resets it to its default state.
     *
     * @return this service filter instance
     */
    DirectoryServiceFilter<T> clear();

    /**
     * Returns the action associated with this service filter.
     * <p>
     * The valid action values depend on the concrete implementation. Most service
     * filters use {@link com.refinitiv.ema.access.FilterEntry.FilterAction} values.
     *
     * @return action
     */
    int action();

    /**
     * Sets the action associated with this service filter.
     * <p>
     * The valid action values depend on the concrete implementation. Most service
     * filters use {@link com.refinitiv.ema.access.FilterEntry.FilterAction} values.
     *
     * @param action the action
     * @return this service filter instance
     * @throws OmmInvalidUsageException if {@code action} is not one of the
     *                                  supported actions
     */
    DirectoryServiceFilter<T> action(int action);

    /**
     * Returns the fixed RDM filter identifier for this service filter type.
     * <p>
     * The returned value corresponds to one of the service filter identifiers defined
     * by {@link EmaRdm}.
     *
     * @return filterId
     */
    int filterId();

    /**
     * Encodes this service filter into its EMA container representation.
     *
     * @return encoded container representing this service filter
     */
    T encode();

    /**
     * Decodes an EMA container representation into this service filter.
     * <p>
     * Implementations typically replace their current content with the decoded data.
     *
     * @param struct encoded container representing the service filter
     * @return this service filter instance
     *
     * @throws OmmInvalidUsageException if {@code struct} is invalid or the filter
     *                                  cannot be decoded
     */
    DirectoryServiceFilter<T> decode(T struct);
}

File: CSharp/Ema/Src/Core/Domain/Directory/DirectoryServiceFilter.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2026 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

using LSEG.Ema.Access;
using System.Text;

namespace LSEG.Ema.Domain.Directory
{
    /// <summary>
    /// The RDM Service Filter. Contains information provided by the Source Directory filter.
    /// </summary>
    /// <typeparam name="T">The type of the complex data associated with this filter.</typeparam>
    /// <typeparam name="TSelf">The type of the derived filter class.</typeparam>
    public abstract class DirectoryServiceFilter<T, TSelf>
        where T : ComplexType
        where TSelf : DirectoryServiceFilter<T, TSelf>
    {
        private readonly StringBuilder m_ToString = new();
        private DirectoryFilterAction m_Action = DirectoryFilterAction.SET;

        /// <summary>
        /// Action associated with this service filter.
        /// </summary>
        /// <returns>The action associated with this service filter.</returns>
        public DirectoryFilterAction Action() => m_Action;
        /// <summary>
        /// Action associated with this service filter.
        /// </summary>
        /// <param name="value">The action to set for this service filter.</param>
        /// <returns>The current instance to support method chaining.</returns>
        public TSelf Action(DirectoryFilterAction value)
        {
            if (!System.Enum.IsDefined(value))
                throw new OmmInvalidUsageException($"Invalid action value: {value}", OmmInvalidUsageException.ErrorCodes.INVALID_ARGUMENT);
            m_Action = value;
            return (TSelf)this;
        }

        /// <summary>
        /// Clears the current contents of the message and prepares it for re-use.
        /// </summary>
        /// <returns>The current instance to support method chaining.</returns>
        public virtual TSelf Clear()
        {
            m_Action = DirectoryFilterAction.SET;
            return (TSelf)this;
        }

        /// <summary>
        /// Copies the state from the specified source instance to the current instance, replacing any existing data.
        /// </summary>
        /// <remarks>This method clears the current instance before copying the state from the source.</remarks>
        /// <param name="source">The instance from which to copy state. Must not be null.</param>
        /// <returns>The current instance to support method chaining.</returns>
        public virtual TSelf CopyFrom(TSelf source)
        {
            Clear();
            m_Action = source.m_Action;
            return (TSelf)this;
        }

        internal abstract void DecodeFrom(T data);
        internal abstract void EncodeTo(T data);

        /// <summary>
        /// Populated by <see cref="Rdm.EmaRdm"/>
        /// </summary>
        public abstract int FilterId { get; }

        /// <inheritdoc />
        public override string ToString()
        {
            m_ToString.Clear();
            AppendToString(m_ToString, 0);
            return m_ToString.ToString();
        }

        internal virtual void AppendToString(StringBuilder sb, int indent)
        {
            sb
                .AddIndent(indent).AppendLine($"FilterId: {FilterId}")
                .AddIndent(indent).AppendLine($"Action: {m_Action}");
        }
    }

    /// <summary>
    /// Represents filter entry action.
    /// </summary>
    public enum DirectoryFilterAction
    {
        /// <summary>
        /// Indicates a partial change of the current Omm data.
        /// </summary>
        UPDATE = FilterAction.UPDATE,
        /// <summary>
        /// Indicates to specify or replace the current Omm data.
        /// </summary>
        SET = FilterAction.SET,
        /// <summary>
        /// Indicates to unset the current Omm data.
        /// </summary>
        CLEAR = FilterAction.CLEAR,
    }
}
